src/signTx.c

```diff
@@ -789,7 +825,70 @@ static void signTx_handleCertificate_ui_runStep()
 
 		advanceCertificatesStateIfAppropriate();
 	}
-	UI_STEP_END(HANDLE_INPUT_STEP_INVALID);
+	UI_STEP_END(HANDLE_CERTIFICATE_STEP_INVALID);
+}
+
+enum {
+	HANDLE_CERTIFICATE_POOL_RETIREMENT_STEP_DISPLAY_OPERATION = 650,
+	HANDLE_CERTIFICATE_POOL_RETIREMENT_STEP_DISPLAY_EPOCH,
+	HANDLE_CERTIFICATE_POOL_RETIREMENT_STEP_CONFIRM,
+	HANDLE_CERTIFICATE_POOL_RETIREMENT_STEP_RESPOND,
+	HANDLE_CERTIFICATE_POOL_RETIREMENT_STEP_INVALID,
+};
+
+static void signTx_handleCertificatePoolRetirement_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	ASSERT(ctx->stageData.certificate.type == CERTIFICATE_TYPE_STAKE_POOL_RETIREMENT);
+	ui_callback_fn_t* this_fn = signTx_handleCertificatePoolRetirement_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step, this_fn);
+
+	UI_STEP(HANDLE_CERTIFICATE_POOL_RETIREMENT_STEP_DISPLAY_OPERATION) {
+		char poolId[200];
+		explicit_bzero(poolId, SIZEOF(poolId));
+
+		size_t length = encode_hex(
```

Comment:
This should be the bech32 pool ID I presume, moreover, now that I'm testing it, I see

```
Retire stake pool
000000000000000000000
```
Do we really need to allocate 200 bytes here? Isn't some stack overflow going on here?

Reply:
I think that the pool key hash fetched from the instruction state is actually empty as it is computed into a local variable when loading it into the tx hash builder:

https://github.com/vacuumlabs/ledger-app-cardano-shelley/blob/01ecd80202758eda1d64901958811feef36081da/src/signTx.c#L984

but then it's fetched from the "poolKeyHash" property which is not passed as part of the certificate data:

https://github.com/vacuumlabs/ledger-app-cardano-shelley/blob/01ecd80202758eda1d64901958811feef36081da/src/signTx.c#L852